src/rgw/rgw_op.h

```diff
@@ -51,6 +51,38 @@ using ceph::crypto::SHA1;
 struct req_state;
 class RGWOp;
 
+// So! Now and then when we try to update bucket information, the
+// bucket has changed during the course of the operation. (Or we have 
+// a cache consistency problem that Watch/Notify isn't ruling out
+// completely.)
+//
+// When this happens, we need to update the bucket info and try 
+// again. We have, however, to try the right *part* again.  We can't
+// simply re-send, since that will obliterate the previous update. 
+//
+// Thus, callers of this function should include everything that
+// merges information to be changed into the bucket information as
+// well as the call to set it. 
+//
+// The called function must return an integer, negative on error. In
+// general, they should just return op_ret.
+//
+// PS: Originally implemented by adamemerson in rgw_op.cc. Just move
+//     it here for more genernal use.
+namespace {
+template<typename F>
+int retry_raced_bucket_write(RGWRados* g, req_state* s, const F& f) { 
+  auto r = f();
+  for (auto i = 0u; i < 15u && r == -ECANCELED; ++i) {
+    r = g->try_refresh_bucket_info(s->bucket_info, nullptr,
+                   &s->bucket_attrs);
+    if (r >= 0) { 
+      r = f();
+    }
+  }
+  return r;
+}
+}
```

Comment:
since this helper is specific to buckets, i think `rgw_bucket.h` is a better home for it. also, since it's being moved into a header file, it shouldn't be wrapped in an anonymous namespace - so either remove that, or use `namespace rgw` instead

Reply:
@cbodley 
So maybe it is better for me to remove the commit 'rgw: Leverage retry_raced_bucket_write when updating quota' from this PR and let's address retry_raced_bucket_write signature improvement in an independent PR for better commit isolation. What do you think?

Reply:
works for me 👍